### PR TITLE
feat: `engine_newPayloadV3`: validate, execute & store block

### LINKED ISSUE #51: Implement `engine_newPayloadV3` (partially) endpoint
Blocked by: #78

See: https://github.com/ethereum/execution-apis/blob/main/src/engine/cancun.md#engine_newpayloadv3

This is the endpoint where we get execution blocks.

### PR DESCRIPTION
**Pull request by fmoletta**
**Motivation**

Being able to fully validate execute and store blocks received by `engine_newPayloadV3`

**Description**

* Add block validation, execution & storing to `engine_newPayloadV3` endpoint

Fixes:
* `Genesis.get_block`: 
  * Now uses the `INITIAL_BASE_FEE` as base_fee_per_gas
  * Now computes roots instead of using zero as default
  * Now uses default values for cancun fields instead of leaving them as None
  
 (With these fixes the genesis' block hash now matches the parentBlockHash of the next block when running with kurtosis)
 * `beacon_root_contract_call` now sets the block's gas_limit to avoid tx validation errors
 
 Misc:
 * `compute_transactions_root` is now a standalone function matching the other compute functions
 * Log more steps of the `engine_newPayloadV3` endpoint
 * Improve api of `ExecutionPayloadV3` & `PayloadStatus`
 
 **Other**
  We can now execute payloads when running with kurtosis 🚀 
Captura de pantalla 2024-08-07 a la(s) 17 38 59

 Disclaimer: We are still getting some execution errors in later blocks that we need to look into (They are all currently passing the block validations)

Closes #51

### CODE REVIEW

File: crates/core/types/genesis.rs [hunk 1]
```diff
@@ -120,19 +122,19 @@ impl Genesis {
...
+            parent_beacon_block_root: Some(H256::zero()),
         }
     }
 
     fn get_block_body(&self) -> BlockBody {
         BlockBody {
             transactions: vec![],
             ommers: vec![],
-            withdrawals: None,
+            withdrawals: Some(vec![]),
```
Comment (mpaulucci) [inline crates/core/types/genesis.rs:137]: Im surprised these changes don't break any of the spec test. Does the rlp and the state root remain equal with these values (empty vec vs None, 0 vs None)?

**Review by mpaulucci — APPROVED** [review]
great work! 🚀

**Comment by mpaulucci** [comment]
Please wait for https://github.com/lambdaclass/ethereum_rust/pull/230 to be fixed

File: crates/core/types/genesis.rs (hunk 1 above)
Comment (fmoletta) [reply crates/core/types/genesis.rs:137]: The state root shouldn't be affected (as it is built from the accounts in the alloc). The rlp does change, and do we need it to change as with the current version block validation failed when running with kurtosis as the genesis block's hash didn't match the parent_block_hash of the following block

File: crates/evm/evm.rs [hunk 2]
```diff
@@ -300,7 +300,7 @@ pub fn beacon_root_contract_call(
     let tx_env = TxEnv {
         caller: *SYSTEM_ADDRESS,
         transact_to: RevmTxKind::Call(*CONTRACT_ADDRESS),
-        gas_limit: 30_000_000,
+        gas_limit: header.gas_limit,
```
Comment (vicentevieytes) [inline crates/evm/evm.rs:303]: This should be set to 30_000_000 according to the EIP 4788 spec:
https://eips.ethereum.org/EIPS/eip-4788#block-processing

Comment (fmoletta) [reply crates/evm/evm.rs:303]: We get the error "caller gas limit exceeds the block gas limit" from the evm with that value. (This comes from the check `self.tx.gas_limit > self.block.gas_limit` in `validate_tx`)

Comment (fmoletta) [reply crates/evm/evm.rs:303]: We could also fix it by setting the block_env's gas_limit to 30_000_000 too

Comment (fmoletta) [reply crates/evm/evm.rs]: Update: This last solution works fine when running with kurtosis, I pushed it

**Pull request merged**